### ISSUE TITLE
GHCJS.libAbiHash includes trailing \n

### ISSUE THREAD

**Issue by stepcut** [haskell/cabal#3915]
I believe that `libAbiHash` from this module:

https://github.com/haskell/cabal/blob/master/Cabal/Distribution/Simple/GHCJS.hs

Is including a trailing newline when calculating the hash. It calculates the hash by calling ghcjs via `getProgramInvocationOutput` and does not appear to do anything to strip any trailing newlines.

In my particular case, I ran into an issue where `libAbiHash` was being used to calculate the `InstalledPackageId` for a reexported-module. The `\n` in the hash was being preserved all the way into the `.conf` file generated by `./Setup register`.

I changed the last line of `libAbiHash` to:

```
  s <- getProgramInvocationOutput verbosity (ghcInvocation ghcjsProg comp ghcArgs)
  return $ filter (/= '\n') s
```

And my problems went away.

I am on OS X. I ran into this issue with `Cabal-1.22.5.0` that is bundled with GHC 7.10.3, but I don't see any reason why it would not still exist.

**Comment by dcoutts** [comment]
We ought to do a vague sanity check on the output. e.g.

    case words s of
      [h] | all isValidHashChar h -> return h
      _ -> fail $ "unexpected output from 'ghcjs --abi-hash': " ++ s

**Comment by ezyang** [comment]
@stepcut Would you like to submit a patch for this? FWIW, here's what GHC does:

```
  hash <- getProgramInvocationOutput verbosity
          (ghcInvocation ghcProg comp platform ghcArgs)
  return (takeWhile (not . isSpace) hash)
```

(@23Skidoo, should we try to get a fix for this in 1.24 patch release? Seems pretty major, although I don't know if a patch release will help GHCJS.)

**Comment by 23Skidoo** [comment]
Yes, would be nice to get a fix into 1.24.
